In this Notebook, I try to analyze whether the RentCast APi will give us relevant information for residential buildings, specifically focusing on the roomCount. Previously, I explorered Zillow and Apartments.com in which the response doest not contain the relevant information.

In [ ]:
# Previous attempt to use zillow/ rapidAPI
# import requests

# url = "https://zillow56.p.rapidapi.com/property"

# property_url ="https://www.zillow.com/homedetails/10811-Termini-San-Luis-Pass-Rd-2112-Galveston-TX-77554/27695234_zpid/"
# querystring = {"zpid":None, "url":property_url}

# headers = {
# 	"X-RapidAPI-Key": "f7bd5fa94cmshb5dade7505aa452p1a684ajsnc91121b96eb2",
# 	"X-RapidAPI-Host": "zillow56.p.rapidapi.com"
    
# }

# response = requests.get(url, headers=headers, params=querystring)

# print(response.json())

This is an example response using the RentCast API

In [132]:
# # Example Response
# [
#   {
#     "id": "5500-Grand-Lake-Dr,-San-Antonio,-TX-78244",
#     "formattedAddress": "5500 Grand Lake Dr, San Antonio, TX 78244",
#     "addressLine1": "5500 Grand Lake Dr",
#     "addressLine2": null,
#     "city": "San Antonio",
#     "state": "TX",
#     "zipCode": "78244",
#     "county": "Bexar",
#     "latitude": 29.475962,
#     "longitude": -98.351442,
#     "propertyType": "Single Family",
#     "bedrooms": 3,
#     "bathrooms": 2,
#     "squareFootage": 1878,
#     "lotSize": 8843,
#     "yearBuilt": 1973,
#     "assessorID": "05076-103-0500",
#     "legalDescription": "CB 5076A BLK 3 LOT 50",
#     "subdivision": "CONV A/S CODE",
#     "zoning": "RH",
#     "lastSaleDate": "2017-10-19T00:00:00.000Z",
#     "lastSalePrice": 185000,
#     "features": {
#       "architectureType": "Contemporary",
#       "cooling": true,
#       "coolingType": "Central",
#       "exteriorType": "Wood",
#       "floorCount": 1,
#       "foundationType": "Slab / Mat / Raft",
#       "garage": true,
#       "garageType": "Garage",
#       "heating": true,
#       "heatingType": "Forced Air",
#       "pool": true,
#       "roofType": "Asphalt",
#       "roomCount": 5,
#       "unitCount": 1
#     },
#     "taxAssessments": {
#       "2019": {
#         "value": 135430,
#         "land": 23450,
#         "improvements": 111980
#       },
#       "2020": {
#         "value": 142610,
#         "land": 23450,
#         "improvements": 119160
#       },
#       "2021": {
#         "value": 163440,
#         "land": 45050,
#         "improvements": 118390
#       },
#       "2022": {
#         "value": 197600,
#         "land": 49560,
#         "improvements": 148040
#       },
#       "2023": {
#         "value": 225790,
#         "land": 59380,
#         "improvements": 166410
#       }
#     },
#     "propertyTaxes": {
#       "2019": {
#         "total": 2984
#       },
#       "2020": {
#         "total": 3023
#       },
#       "2021": {
#         "total": 3455
#       },
#       "2022": {
#         "total": 4091
#       },
#       "2023": {
#         "total": 4675
#       }
#     },
#     "owner": {
#       "names": [
#         "Michael Smith"
#       ],
#       "mailingAddress": {
#         "id": "149-Weaver-Blvd,---264,-Weaverville,-NC-28787",
#         "formattedAddress": "149 Weaver Blvd, # 264, Weaverville, NC 28787",
#         "addressLine1": "149 Weaver Blvd",
#         "addressLine2": "# 264",
#         "city": "Weaverville",
#         "state": "NC",
#         "zipCode": "28787"
#       }
#     },
#     "ownerOccupied": false
#   }
# ]

### I defined two functions to help me see the relevant information

In [188]:
def check_roomCounts(data):
    """
    Prints the formatted address and room count for each property in the data list that contains a room count.
    
    Iterates through a list of property dictionaries, checks for the presence of a room count in the property's features,
    and prints out the property's formatted address alongside the room count if it exists. Additionally, it counts
    the number of properties that contain a room count.
    
    Parameters:
    - data (list of dict): A list of dictionaries, where each dictionary represents a property and contains
      a 'formattedAddress' key and a 'features' dictionary that may include a 'roomCount' key.
    
    Returns:
    - int: The number of properties that contain a room count.
    """
    count_contained = 0
    for item in data:
        formattedAddress = item.get('formattedAddress')
        features = item.get('features',{})
        roomCount = features.get('roomCount', None)
        if roomCount is not None:
            print(f"Address: {formattedAddress}, Room Count: {roomCount}")
            count_contained += 1
    return count_contained

In [189]:
def check_features(data):
    """
    Prints the features of each property in the data list that contains additional feature information.
    
    Iterates through a list of property dictionaries and prints the 'features' dictionary for each property,
    if it exists. The 'features' dictionary is assumed to contain additional details about the property.
    
    Parameters:
    - data (list of dict): A list of dictionaries, where each dictionary represents a property and
      may contain a 'features' dictionary with additional property details.
    
    Returns:
    - None
    """
    for item in data:
        features = item.get('features', {})  # Returns {} if 'features' is not found
        if features:  # This will be False if features is {}
            print(features)


### PART 1: Exploring the RentCast API by trying out different types of properties
propertyType (string):
The type of the property, used to search for properties matching this criteria. 
https://developers.rentcast.io/reference/property-types

In [77]:
# For Apartments in Ann Arbor:
import requests

url_Ann_Apartments = "https://api.rentcast.io/v1/properties?city=Ann%20Arbor&state=MI&propertyType=Apartment&limit=500&offset=0"

headers_Ann_Apartments = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_Ann_Apartments = requests.get(url_Ann_Apartments, headers=headers_Ann_Apartments).json()

In [201]:
check_roomCounts(response_Ann_Apartments)
check_features(response_Ann_Apartments)

Address: 437 2nd St, Ann Arbor, MI 48103, Room Count: 12
{'architectureType': 'Condo / Apartment', 'heating': True, 'heatingType': 'Forced Air', 'unitCount': 1}
{'cooling': True, 'coolingType': 'Package', 'heating': True, 'heatingType': 'Other', 'architectureType': 'Condo / Apartment'}
{'architectureType': 'Contemporary'}
{'heating': True, 'heatingType': 'Hot Water', 'architectureType': 'Condo / Apartment'}
{'fireplace': True, 'foundationType': 'Brick', 'heating': True, 'heatingType': 'Forced Air', 'roomCount': 12, 'unitCount': 1}
500


In [203]:
print(len(response_Ann_Apartments))

500


### Since each API response contains maximum 500 properties information, we need to utilize the setoff QUERY PARAM if we want to get a more complete dataset.

In [59]:
# Apartment information for Savannah, GA

import requests

url_GA_Apa= "https://api.rentcast.io/v1/properties?city=Savannah&state=GA&propertyType=Apartment&limit=500&offset=0"

headers_GA_Apa = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_GA_Apa = requests.get(url_GA_Apa, headers=headers_GA_Apa).json()

In [202]:
check_roomCounts(response_GA_Apa)
check_features(response_GA_Apa)

Address: 403 W 34th St, Savannah, GA 31401, Room Count: 6
Address: 402 W 33rd St, Savannah, GA 31401, Room Count: 10
Address: 1805 Martin Luther King Jr Blvd, Savannah, GA 31415, Room Count: 10
Address: 606 E Park Avenue Ln, Savannah, GA 31401, Room Count: 16
Address: 1302 E Victory Dr, Apt 1, Savannah, GA 31404, Room Count: 20
Address: 2511 Price St, Apt 3, Savannah, GA 31401, Room Count: 20
Address: 301 W Victory Dr, Savannah, GA 31405, Room Count: 4
Address: 809 W 35th St, Savannah, GA 31415, Room Count: 20
Address: 9844 Whitefield Ave, Savannah, GA 31406, Room Count: 4
Address: 641 E Anderson St, Savannah, GA 31401, Room Count: 8
Address: 1011 E 39th St, Savannah, GA 31401, Room Count: 6
Address: 537 E 37th St, Savannah, GA 31401, Room Count: 7
Address: 1403 E 51st St, Savannah, GA 31404, Room Count: 7
Address: 511 W 37th St, Savannah, GA 31415, Room Count: 6
Address: 9236 Garland Dr, Savannah, GA 31406, Room Count: 5
{'architectureType': 'Condo / Apartment', 'floorCount': 2, 'exte

In [204]:
print(len(response_GA_Apa))

500


In [65]:
# Condo in Savannah, GA:
import requests

url_Savannah_condo = "https://api.rentcast.io/v1/properties?city=Savannah&state=GA&propertyType=Condo&limit=500&offset=0"

headers_Savannah_condo = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_Savannah_condo = requests.get(url_Savannah_condo, headers=headers_Savannah_condo).json()

In [129]:
check_roomCounts(response_Savannah_condo)
check_features(response_Savannah_condo)

Address: 211 W 60th St, Savannah, GA 31405, Room Count: 5
{'exteriorType': 'Brick Veneer'}
{'exteriorType': 'Brick Veneer'}
{'architectureType': 'Condo / Apartment'}
{'exteriorType': 'Metal'}
{'exteriorType': 'Concrete Block'}
{'architectureType': 'Condo / Apartment'}
{'exteriorType': 'Stucco'}
{'exteriorType': 'Concrete Block'}
{'exteriorType': 'Stucco'}
{'exteriorType': 'Aluminum / Vinyl Siding'}
{'exteriorType': 'Concrete Block'}
{'exteriorType': 'Aluminum / Vinyl Siding'}
{'exteriorType': 'Concrete Block'}
{'exteriorType': 'Concrete Block'}
{'exteriorType': 'Brick'}
{'exteriorType': 'Stucco'}
{'exteriorType': 'Brick Veneer'}
{'exteriorType': 'Brick Veneer'}
{'exteriorType': 'Concrete Block'}
{'architectureType': 'Condo / Apartment'}
{'exteriorType': 'Brick'}
{'exteriorType': 'Concrete Block'}
{'exteriorType': 'Brick'}
{'exteriorType': 'Brick Veneer'}
{'exteriorType': 'Brick Veneer'}
{'exteriorType': 'Brick'}
{'exteriorType': 'Stucco'}
{'exteriorType': 'Brick'}
{'exteriorType': 'Bri

In [14]:
# Apartment information for Hilo
import requests

url = "https://api.rentcast.io/v1/properties?city=Hilo&state=HI&propertyType=Apartment&limit=500&offset=0"

headers = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_adj = requests.get(url, headers=headers).json()
# response_adj

[{'id': '189-Kukuau-St,-Unit-100,-Hilo,-HI-96720',
  'formattedAddress': '189 Kukuau St, Unit 100, Hilo, HI 96720',
  'addressLine1': '189 Kukuau St',
  'addressLine2': 'Unit 100',
  'city': 'Hilo',
  'state': 'HI',
  'zipCode': '96720',
  'county': 'Hawaii',
  'latitude': 19.718335,
  'longitude': -155.085895,
  'propertyType': 'Apartment',
  'squareFootage': 3276,
  'lotSize': 9441,
  'yearBuilt': 2002,
  'assessorID': '3230180370000',
  'legalDescription': 'LOT 2 9441 SF MAP 2 LCAPP 885',
  'subdivision': 'MAP 02',
  'zoning': '3A',
  'features': {'architectureType': 'Condo / Apartment'},
  'taxAssessments': {'2018': {'value': 566500,
    'land': 169900,
    'improvements': 396600}},
  'propertyTaxes': {'2017': {'total': 3515}},
  'owner': {'names': ['KUKUAU COTTAGE LLC'],
   'mailingAddress': {'id': '189-Kukuau-St,-UNIT-100,-Hilo,-HI-96720',
    'addressLine1': '189 Kukuau St',
    'addressLine2': 'UNIT 100',
    'city': 'Hilo',
    'state': 'HI',
    'zipCode': '96720'}},
  'owner

In [102]:
check_roomCounts(response_adj)
check_features(response_adj)

Address: 234 Kapiolani St, Hilo, HI 96720, Room Count: 4
Address: 310 Ululani St, Unit C, Hilo, HI 96720, Room Count: 5
Address: 265 Kekuanaoa St, # D, Hilo, HI 96720, Room Count: 8
Address: 237 Kaumana Dr, Hilo, HI 96720, Room Count: 5
Address: 282 Kukuau St, Unit C, Hilo, HI 96720, Room Count: 24
{'architectureType': 'Condo / Apartment'}
{'exteriorType': 'Metal', 'floorCount': 3, 'unitCount': 1}
{'exteriorType': 'Wood', 'floorCount': 1, 'unitCount': 1}
{'architectureType': 'Condo / Apartment', 'unitCount': 1}
{'exteriorType': 'Wood', 'floorCount': 2, 'unitCount': 1}
{'exteriorType': 'Wood', 'floorCount': 2, 'unitCount': 1}
{'architectureType': 'Condo / Apartment'}
{'exteriorType': 'Masonry', 'floorCount': 1, 'unitCount': 1}
{'architectureType': 'Condo / Apartment'}
{'architectureType': 'Condo / Apartment', 'exteriorType': 'Wood', 'floorCount': 2, 'unitCount': 4}
{'architectureType': 'Condo / Apartment', 'floorCount': 1, 'unitCount': 1}
{'architectureType': 'Condo / Apartment', 'floor

In [205]:
len(response_adj)

101

In [50]:
# Townhouses in Hilo:
import requests

url = "https://api.rentcast.io/v1/properties?city=Hilo&state=HI&propertyType=Townhouse&limit=500&offset=0"

headers = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_townhouse_HI = requests.get(url, headers=headers).json()

In [107]:
# For the townhouse category, we see that the feature column is empty.
check_roomCounts(response_townhouse_HI)
#check_features(response_townhouse_HI)

In [108]:
# Multi-family information for Ann Arbor, MI
import requests

url_Ann_Multi = "https://api.rentcast.io/v1/properties?city=Ann%20Arbor&state=MI&propertyType=Multi-Family&limit=500&offset=0"

headers_Ann_Multi = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_Ann_Multi = requests.get(url_Ann_Multi, headers=headers_Ann_Multi).json()


In [110]:
check_roomCounts(response_Ann_Multi)
check_features(response_Ann_Multi)

Address: 426 3rd St, Ann Arbor, MI 48103, Room Count: 7
Address: 2130 Tremmel Ave, Ann Arbor, MI 48104, Room Count: 4
Address: 401 Pauline Blvd, Ann Arbor, MI 48103, Room Count: 6
Address: 1619 Long Meadow Trl, Ann Arbor, MI 48108, Room Count: 4
Address: 1686 Coburn Dr, Ann Arbor, MI 48108, Room Count: 4
Address: 1632 Coburn Dr, Ann Arbor, MI 48108, Room Count: 4
Address: 1707 Coburn Ct, Ann Arbor, MI 48108, Room Count: 4
Address: 1714 Charlton Ave, Ann Arbor, MI 48103, Room Count: 8
Address: 1443 Jewett St, Ann Arbor, MI 48104, Room Count: 8
Address: 1705 Coburn Ct, Ann Arbor, MI 48108, Room Count: 4
Address: 1608 Oakfield Dr, Ann Arbor, MI 48108, Room Count: 4
Address: 1656 Coburn Dr, Ann Arbor, MI 48108, Room Count: 4
Address: 1621 Long Meadow Trl, Ann Arbor, MI 48108, Room Count: 4
Address: 1678 Weatherstone Dr, Ann Arbor, MI 48108, Room Count: 4
Address: 1736 Weatherstone Dr, Ann Arbor, MI 48108, Room Count: 5
Address: 1666 Coburn Dr, Ann Arbor, MI 48108, Room Count: 5
Address: 15

In [112]:
# We see that multi-family could be important to consider.
# Multi-family information for Hilo, Hi:
import requests

url_Hilo_Multi = "https://api.rentcast.io/v1/properties?city=Hilo&state=HI&propertyType=Multi-Family&limit=500&offset=0"

headers_Hilo_Multi = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_Hilo_Multi = requests.get(url_Hilo_Multi, headers=headers_Hilo_Multi).json()

In [114]:
check_roomCounts(response_Hilo_Multi)
check_features(response_Hilo_Multi)

{'floorCount': 1, 'garage': True, 'unitCount': 3}


# Galveston, Texas
## 1. Multi-family

In [115]:
import requests

url_TX_Multi = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Multi-Family&limit=500&offset=0"

headers_TX_Multi = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Multi = requests.get(url_TX_Multi, headers=headers_TX_Multi).json()

In [269]:
check_roomCounts(response_TX_Multi )
check_features(response_TX_Multi )
print(len(response_TX_Multi))

Address: 1015 19th St, Galveston, TX 77550, Room Count: 8
Address: 902 Church St, Galveston, TX 77550, Room Count: 10
Address: 1902 27th St, Galveston, TX 77550, Room Count: 10
Address: 2116 25th St, Galveston, TX 77550, Room Count: 10
Address: 2114 Sealy St, Galveston, TX 77550, Room Count: 18
Address: 1721 22nd St, Galveston, TX 77550, Room Count: 2
Address: 811 7th St, Galveston, TX 77550, Room Count: 2
Address: 2524 Ave K, Galveston, TX 77550, Room Count: 2
Address: 1117 31st St, Galveston, TX 77550, Room Count: 6
Address: 2302 54th St, Galveston, TX 77551, Room Count: 4
Address: 909 18th St, Galveston, TX 77550, Room Count: 2
Address: 1606 23rd St, Galveston, TX 77550, Room Count: 2
Address: 517 University Blvd, Galveston, TX 77550, Room Count: 8
Address: 2012 57th St, Galveston, TX 77551, Room Count: 3
Address: 1128 Winnie St, Galveston, TX 77550, Room Count: 2
Address: 1912 Ave K, Galveston, TX 77550, Room Count: 2
Address: 3002 Ave R, Galveston, TX 77550, Room Count: 8
Address:

In [241]:
import requests

url_TX_Multi_2 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Multi-Family&limit=500&offset=501"

headers_TX_Multi_2 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Multi_2 = requests.get(url_TX_Multi_2, headers=headers_TX_Multi_2).json()


In [242]:
len(response_TX_Multi_2)
# in total, we have 500+372=872 multi-family type properties.

372

# Galveston, Texas
## 2. Manufactured

In [123]:
# Texas Manufactured category
import requests

url_TX_Manu = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Manufactured&limit=500&offset=0"

headers_TX_Manu = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Manu = requests.get(url_TX_Manu, headers=headers_TX_Manu).json()

In [239]:
check_roomCounts(response_TX_Manu)
check_features(response_TX_Manu)
print(len(response_TX_Manu))
# Not a lot under the Manufactured category

{'architectureType': 'Mobile Home', 'floorCount': 1, 'garage': True, 'garageSpaces': 1, 'garageType': 'Carport', 'unitCount': 1}
{'floorCount': 1, 'garage': True, 'garageType': 'Detached', 'unitCount': 1}
{'floorCount': 1}
{'floorCount': 1}
{'floorCount': 1, 'unitCount': 1}
{'architectureType': 'Conventional', 'cooling': True, 'coolingType': 'Central', 'exteriorType': 'Wood Frame / Siding', 'floorCount': 2, 'foundationType': 'Pile', 'garage': True, 'garageType': 'Carport', 'heating': True, 'heatingType': 'Central', 'roofType': 'Composition Shingle'}
{'architectureType': 'Mobile Home', 'cooling': True, 'coolingType': 'Central', 'foundationType': 'Pile', 'garage': True, 'garageType': 'Mixed', 'heating': True, 'heatingType': 'Central', 'roofType': 'Composition Shingle'}
7


# Galveston, Texas
## 3. Land

In [126]:
# Texas Land category
import requests

url_TX_Land = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Land&limit=500&offset=0"

headers_TX_Land = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Land = requests.get(url_TX_Land, headers=headers_TX_Land).json()

In [127]:
check_roomCounts(response_TX_Land)
check_features(response_TX_Land)

{'floorCount': 1, 'unitCount': 1}
{'architectureType': 'Traditional', 'cooling': True, 'coolingType': 'Commercial', 'exteriorType': 'Wood Frame / Siding', 'floorCount': 2, 'foundationType': 'Pier', 'garage': True, 'garageType': 'Carport', 'heating': True, 'heatingType': 'Floor Furnace', 'roofType': 'Gravel', 'unitCount': 1}


# Galveston, Texas
## 4. Condo

In [130]:
# Texas Condo category
import requests

url_TX_Condo = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Condo&limit=500&offset=0"

headers_TX_Condo = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Condo = requests.get(url_TX_Condo, headers=headers_TX_Condo).json()

In [131]:
check_roomCounts(response_TX_Condo)
check_features(response_TX_Condo)

{'architectureType': 'Condo / Apartment', 'floorCount': 1, 'unitCount': 1}
{'architectureType': 'Condo / Apartment', 'floorCount': 1, 'unitCount': 1}
{'architectureType': 'Condo / Apartment', 'floorCount': 1, 'unitCount': 1}
{'architectureType': 'Condo / Apartment', 'cooling': True, 'coolingType': 'Central', 'floorCount': 1, 'foundationType': 'Slab', 'heating': True, 'heatingType': 'Central', 'unitCount': 1}
{'architectureType': 'Condo / Apartment', 'floorCount': 1, 'unitCount': 1}
{'architectureType': 'Condo / Apartment', 'cooling': True, 'coolingType': 'Central', 'floorCount': 1, 'foundationType': 'Slab', 'heating': True, 'heatingType': 'Central', 'unitCount': 1}
{'architectureType': 'Condo / Apartment', 'cooling': True, 'coolingType': 'Central', 'floorCount': 1, 'foundationType': 'Slab', 'heating': True, 'heatingType': 'Central', 'unitCount': 1}
{'architectureType': 'Condo / Apartment', 'floorCount': 1, 'unitCount': 1}
{'architectureType': 'Condo / Apartment', 'floorCount': 1, 'unit

In [225]:
len(response_TX_Condo)
# Offset!
# Texas Condo category
import requests

url_TX_Condo_2 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Condo&limit=500&offset=501"

headers_TX_Condo_2 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Condo_2 = requests.get(url_TX_Condo_2, headers=headers_TX_Condo_2).json()


In [228]:
len(response_TX_Condo_2)
# 500, offset=1001
import requests

url_TX_Condo_3 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Condo&limit=500&offset=1001"

headers_TX_Condo_3 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Condo_3 = requests.get(url_TX_Condo_3, headers=headers_TX_Condo_3).json()

In [229]:
len(response_TX_Condo_3)

500

In [230]:
import requests

url_TX_Condo_4 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Condo&limit=500&offset=1501"

headers_TX_Condo_4 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Condo_4 = requests.get(url_TX_Condo_4, headers=headers_TX_Condo_4).json()

In [231]:
len(response_TX_Condo_4)

500

In [232]:
import requests

url_TX_Condo_5 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Condo&limit=500&offset=2001"

headers_TX_Condo_5 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Condo_5 = requests.get(url_TX_Condo_5, headers=headers_TX_Condo_5).json()

In [233]:
len(response_TX_Condo_5)

500

In [234]:
import requests

url_TX_Condo_6 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Condo&limit=500&offset=2501"

headers_TX_Condo_6 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Condo_6 = requests.get(url_TX_Condo_6, headers=headers_TX_Condo_6).json()


In [235]:
len(response_TX_Condo_6)

500

In [236]:
import requests

url_TX_Condo_7 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Condo&limit=500&offset=3001"

headers_TX_Condo_7 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Condo_7 = requests.get(url_TX_Condo_7, headers=headers_TX_Condo_7).json()


In [237]:
len(response_TX_Condo_7)

37

### We see that there are a lot of properties under the condo category!

# Galveston, Texas
## 5. Single Family

In [246]:
import requests

url_TX_single = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Single%20Family&limit=500&offset=0"

headers_TX_single = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_single = requests.get(url_TX_single, headers=headers_TX_single).json()


In [248]:
len(response_TX_single)

500

In [250]:
import requests

url_TX_single_2 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Single%20Family&limit=500&offset=501"

headers_TX_single_2 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_single_2 = requests.get(url_TX_single_2, headers=headers_TX_single_2).json()


In [251]:
len(response_TX_single_2)

500

In [253]:
import requests

url_TX_single_3 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Single%20Family&limit=500&offset=1001"

headers_TX_single_3 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_single_3 = requests.get(url_TX_single_3, headers=headers_TX_single_3).json()


In [254]:
len(response_TX_single_3)

500

In [255]:
import requests

url_TX_single_4 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Single%20Family&limit=500&offset=1501"

headers_TX_single_4 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_single_4 = requests.get(url_TX_single_4, headers=headers_TX_single_4).json()


In [256]:
len(response_TX_single_4)

500

In [257]:
import requests

url_TX_single_5 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Single%20Family&limit=500&offset=2001"

headers_TX_single_5 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_single_5 = requests.get(url_TX_single_5, headers=headers_TX_single_5).json()


In [258]:
len(response_TX_single_5)

500

In [259]:
import requests

url_TX_single_6 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Single%20Family&limit=500&offset=2501"

headers_TX_single_6 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_single_6 = requests.get(url_TX_single_6, headers=headers_TX_single_6).json()


In [261]:
len(response_TX_single_6)

500

In [263]:
import requests

url_TX_single_7 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Single%20Family&limit=500&offset=2501"

headers_TX_single_7 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_single_7 = requests.get(url_TX_single_7, headers=headers_TX_single_7).json()


In [264]:
len(response_TX_single_7)

500

In [265]:
import requests

url_TX_single_8 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Single%20Family&limit=500&offset=3001"

headers_TX_single_8 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_single_8 = requests.get(url_TX_single_8, headers=headers_TX_single_8).json()


In [266]:
len(response_TX_single_8)

500

In [267]:
import requests

url_TX_single_9 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Single%20Family&limit=500&offset=3501"

headers_TX_single_9 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_single_9 = requests.get(url_TX_single_9, headers=headers_TX_single_9).json()


In [268]:
len(response_TX_single_9)

500

# Galveston, Texas
## 6. Apartment

In [139]:
import requests

url_TX_Apar = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Apartment&limit=500&offset=0"

headers_TX_Apar = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Apar = requests.get(url_TX_Apar, headers=headers_TX_Apar).json()

In [149]:
response_TX_Apar

[{'id': '4400-Ave-N,-Galveston,-TX-77550',
  'formattedAddress': '4400 Ave N, Galveston, TX 77550',
  'addressLine1': '4400 Ave N',
  'addressLine2': None,
  'city': 'Galveston',
  'state': 'TX',
  'zipCode': '77550',
  'county': 'Galveston',
  'latitude': 29.290238,
  'longitude': -94.814662,
  'propertyType': 'Apartment',
  'yearBuilt': 1985,
  'lastSaleDate': '2011-01-07T00:00:00.000Z'},
 {'id': '2101-Church-Rear-St,-Galveston,-TX-77550',
  'formattedAddress': '2101 Church Rear St, Galveston, TX 77550',
  'addressLine1': '2101 Church Rear St',
  'addressLine2': None,
  'city': 'Galveston',
  'state': 'TX',
  'zipCode': '77550',
  'county': 'Galveston County',
  'latitude': 29.30409,
  'longitude': -94.79115,
  'propertyType': 'Apartment',
  'yearBuilt': 1927},
 {'id': '7302-Ave-P,-1-2-Ave,-Galveston,-TX-77551',
  'formattedAddress': '7302 Ave P, 1/2 Ave, Galveston, TX 77551',
  'addressLine1': '7302 Ave P',
  'addressLine2': '1/2 Ave',
  'city': 'Galveston',
  'state': 'TX',
  'zipC

In [163]:
len(response_TX_Apar)
# We need to call the API again, use offset to get more apartments

500

In [145]:
import requests

url_TX_Apar_2 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Apartment&limit=500&offset=501"

headers_TX_Apar_2 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Apar_2 = requests.get(url_TX_Apar_2, headers=headers_TX_Apar_2).json()

In [162]:
len(response_TX_Apar_2)
#response_TX_Apar_2

500

In [150]:
import requests

url_TX_Apar_3 = "https://api.rentcast.io/v1/properties?city=Galveston&state=TX&propertyType=Apartment&limit=500&offset=1001"

headers_TX_Apar_3 = {
    "accept": "application/json",
    "X-Api-Key": "889bfd9166f14faa853a08628f388dbb"
}

response_TX_Apar_3 = requests.get(url_TX_Apar_3, headers=headers_TX_Apar_3).json()

In [152]:
len(url_TX_Apar_3)
# We have exausted all the apartments in Galveston, TX
# We now merge the three Json data objects into one

106

In [180]:
# Merging url_TX_Apar, url_TX_Apar_2, and url_TX_Apar_3
combined_json_TX_apar = response_TX_Apar+response_TX_Apar_2+response_TX_Apar_3
total_apart_tx = len(combined_json_TX_apar)
print(f"Total number of Apartments available for Galveston,TX is:{total_apart_tx}")

Total number of Apartments available for Galveston,TX is:1246


In [193]:
# Analyze 
check_roomCounts(combined_json_TX_apar)

Address: 3421 Ave N, Galveston, TX 77550, Room Count: 10
Address: 1302 39th St, Galveston, TX 77550, Room Count: 10
Address: 2802 Winnie St, Galveston, TX 77550, Room Count: 9
Address: 3114 Seawall Blvd, Galveston, TX 77550, Room Count: 8
Address: 2728 Ave Q, Galveston, TX 77550, Room Count: 1
Address: 2805 Ave M, Galveston, TX 77550, Room Count: 8
Address: 16602 Termini San Luis Pass Rd, Galveston, TX 77554, Room Count: 8
Address: 7114 N Holiday Dr, Galveston, TX 77550, Room Count: 8
Address: 7093 N Holiday Dr, Galveston, TX 77550, Room Count: 8
Address: 3426 Ave S, Galveston, TX 77550, Room Count: 10
Address: 7032 N Holiday Dr, Galveston, TX 77550, Room Count: 8
Address: 3916 Broadway St, Galveston, TX 77550, Room Count: 12
Address: 1004 Broadway St, Galveston, TX 77550, Room Count: 10
Address: 1401 Bayou Shore Dr, Galveston, TX 77551, Room Count: 10
Address: 713 Ball St, Galveston, TX 77550, Room Count: 8
Address: 5615 Ave R, Galveston, TX 77551, Room Count: 8
Address: 2308 53rd St,

26

In [196]:
# Determine roughly how many times building unit counts are left blank or null.
number_contains_roomCount_TX = check_roomCounts(combined_json_TX_apar)
# Ratio of buildings that are left blank or null
ratio_TX_apart = number_contains_roomCount_TX / total_apart_tx

Address: 3421 Ave N, Galveston, TX 77550, Room Count: 10
Address: 1302 39th St, Galveston, TX 77550, Room Count: 10
Address: 2802 Winnie St, Galveston, TX 77550, Room Count: 9
Address: 3114 Seawall Blvd, Galveston, TX 77550, Room Count: 8
Address: 2728 Ave Q, Galveston, TX 77550, Room Count: 1
Address: 2805 Ave M, Galveston, TX 77550, Room Count: 8
Address: 16602 Termini San Luis Pass Rd, Galveston, TX 77554, Room Count: 8
Address: 7114 N Holiday Dr, Galveston, TX 77550, Room Count: 8
Address: 7093 N Holiday Dr, Galveston, TX 77550, Room Count: 8
Address: 3426 Ave S, Galveston, TX 77550, Room Count: 10
Address: 7032 N Holiday Dr, Galveston, TX 77550, Room Count: 8
Address: 3916 Broadway St, Galveston, TX 77550, Room Count: 12
Address: 1004 Broadway St, Galveston, TX 77550, Room Count: 10
Address: 1401 Bayou Shore Dr, Galveston, TX 77551, Room Count: 10
Address: 713 Ball St, Galveston, TX 77550, Room Count: 8
Address: 5615 Ave R, Galveston, TX 77551, Room Count: 8
Address: 2308 53rd St,

In [197]:
print(f"Number of properties with null or blank 'unitCount': {ratio_TX_apart}")

Number of properties with null or blank 'unitCount': 0.02086677367576244


We see that around 2% of the apartments in Galveston, TX contains roomCount information.

In [206]:
combined_json_TX_apar

[{'id': '4400-Ave-N,-Galveston,-TX-77550',
  'formattedAddress': '4400 Ave N, Galveston, TX 77550',
  'addressLine1': '4400 Ave N',
  'addressLine2': None,
  'city': 'Galveston',
  'state': 'TX',
  'zipCode': '77550',
  'county': 'Galveston',
  'latitude': 29.290238,
  'longitude': -94.814662,
  'propertyType': 'Apartment',
  'yearBuilt': 1985,
  'lastSaleDate': '2011-01-07T00:00:00.000Z'},
 {'id': '2101-Church-Rear-St,-Galveston,-TX-77550',
  'formattedAddress': '2101 Church Rear St, Galveston, TX 77550',
  'addressLine1': '2101 Church Rear St',
  'addressLine2': None,
  'city': 'Galveston',
  'state': 'TX',
  'zipCode': '77550',
  'county': 'Galveston County',
  'latitude': 29.30409,
  'longitude': -94.79115,
  'propertyType': 'Apartment',
  'yearBuilt': 1927},
 {'id': '7302-Ave-P,-1-2-Ave,-Galveston,-TX-77551',
  'formattedAddress': '7302 Ave P, 1/2 Ave, Galveston, TX 77551',
  'addressLine1': '7302 Ave P',
  'addressLine2': '1/2 Ave',
  'city': 'Galveston',
  'state': 'TX',
  'zipC

### Read the Padlet for Galveston, TX

In [211]:
import pandas as pd

df_excel = pd.read_excel('/Users/kd6801/Desktop/padlet-TX.xlsx')
# Display the first few rows of the dataframe
df_excel.head()

Post number       Section                     Subject  \
0            1  Outdoor Bins               Leeland House   
1            2  Outdoor Bins                 Chick-fil-A   
2            3  Outdoor Bins               Peking Buffet   
3            4  Outdoor Bins  Henry's Mexican Restaurant   
4            5  Outdoor Bins              Blake's Bistro   

                                                Body  \
0              Source: yelp\nCategories: Restaurants   
1  Source: yelp\nCategories: Fast Food|Chicken Sh...   
2  Source: yelp\nCategories: Chinese|Buffets|Asia...   
3  Source: yelp\nCategories: Mexican|Seafood|Burgers   
4  Source: yelp\nCategories: American|Breakfast &...   

                                          Attachment Attachment caption  \
0  https://www.yelp.com/biz/leeland-house-galvest...        Website Url   
1  https://www.yelp.com/biz/chick-fil-a-galveston...        Website Url   
2  https://www.yelp.com/biz/peking-buffet-galvest...        Website Url   
3  https://www.yelp.com/biz/henrys-mexican-restau...        Website Url   
4  https://www.yelp.com/biz/blakes-bistro-galvest...        Website Url   

                                      Location   Latitude  Longitude  \
0          2711 Market St, Galveston, TX 77550  29.303881 -94.798849   
1       2501 Seawall Blvd, Galveston, TX 77550  29.288736 -94.790329   
2  6125 Central City Blvd, Galveston, TX 77551  29.268920 -94.828520   
3        2823 Broadway St, Galveston, TX 77550  29.297983 -94.798500   
4       5222 Seawall Blvd, Galveston, TX 77551  29.272540 -94.818292   

                     Author               Created At               Updated At  
0  Launa Greer (launagreer)  2024-03-29 21:15:17 UTC  2024-03-29 21:15:17 UTC  
1  Launa Greer (launagreer)  2024-03-29 21:15:18 UTC  2024-03-29 21:15:18 UTC  
2  Launa Greer (launagreer)  2024-03-29 21:15:18 UTC  2024-03-29 21:15:18 UTC  
3  Launa Greer (launagreer)  2024-03-29 21:15:19 UTC  2024-03-29 21:15:19 UTC  
4  Launa Greer (launagreer)  2024-03-29 21:15:21 UTC  2024-03-29 21:15:21 UTC

In [220]:
# Create a set of locations from the 'Location' column
padlet_locations_TX = set(df_excel['Location'].unique())
#padlet_locations_TX
len(padlet_locations_TX)

249

In [221]:
# Create a set of formatted addresses from API data
api_addresses = {item['formattedAddress'] for item in combined_json_TX_apar}
#api_addresses
len(api_addresses)

1246

In [219]:
# Find matches between API addresses and Padlet locations
matches = api_addresses.intersection(padlet_locations_TX)

# Count and print the number of matches
print(f"Number of matching locations: {len(matches)}")


Number of matching locations: 17


#### We have 249 Outdoor Bins locations in the padlet excel file, and we see that Number of matching locations is 17. Maybe we can combine all the API Response data for all types and again figure out the number of matching locations. 

In [277]:
# Combined all categories
TX_final_combined = combined_json_TX_apar + response_TX_single + response_TX_single_2 + response_TX_single_3 + response_TX_single_4 + response_TX_single_5 + response_TX_single_6 + response_TX_single_7 + response_TX_single_8 + response_TX_single_9+response_TX_Condo + response_TX_Condo_2 + response_TX_Condo_3 + response_TX_Condo_4 + response_TX_Condo_5 + response_TX_Condo_6 + response_TX_Condo_7+response_TX_Land+response_TX_Manu+response_TX_Multi+response_TX_Multi_2

In [278]:
print(len(TX_final_combined))
# We now have 10162 entries

10162


In [283]:
# Create a set of formatted addresses from API data
api_addresses_combined = {item['formattedAddress'] for item in TX_final_combined}
# See how many of these we have
len(api_addresses_combined)

9662

In [282]:
# Find matches between API addresses and Padlet locations
matches_2 = api_addresses_combined.intersection(padlet_locations_TX_combined)

# Count and print the number of matches
print(f"Number of matching locations, from a larger API dataset: {len(matches_2)}")


Number of matching locations, from a larger API dataset: 27


We see that our number matching location increases from 17 to 27.

In [285]:
# The number of entries/properties classified as Apartments/Condo:
apartment_condo_entries = df_excel['Body'].str.contains('Apartments|Condominiums', case=False, na=False)
# Count the number of true entries in the filtered result
count = apartment_condo_entries.sum()

print(f"Number of entries for Apartments or Condominiums: {count}")

Number of entries for Apartments or Condominiums: 30


### We see that 27 out of 30 locations are covered by the RentCast API. For Galveston, TX, the coverage rate is 27/30=90%.

In [286]:
# Here are the matched locations
matches_2

{'1108 Ave L, Galveston, TX 77550',
 '1302 39th St, Galveston, TX 77550',
 '1524 Strand St, Galveston, TX 77550',
 '2016 Strand St, Galveston, TX 77550',
 '2027 25th St, Galveston, TX 77550',
 '210 Strand St, Galveston, TX 77550',
 '215 Market St, Galveston, TX 77550',
 '2208 Post Office St, Galveston, TX 77550',
 '2415 Ave N, Galveston, TX 77550',
 '2915 Ball St, Galveston, TX 77550',
 '302 25th St, Galveston, TX 77550',
 '3102 Cove View Blvd, Galveston, TX 77554',
 '3219 69th St, Galveston, TX 77551',
 '3222 69th St, Galveston, TX 77551',
 '3428 Cove View Blvd, Galveston, TX 77554',
 '3433 Cove View Blvd, Galveston, TX 77554',
 '3506 Cove View Blvd, Galveston, TX 77554',
 '4400 Ave N, Galveston, TX 77550',
 '4925 Fort Crockett Blvd, Galveston, TX 77551',
 '6102 Seawall Blvd, Galveston, TX 77551',
 '6300 Seawall Blvd, Galveston, TX 77551',
 '6424 Central City Blvd, Galveston, TX 77551',
 '7019 Lasker Dr, Galveston, TX 77551',
 '7310 Seawall Blvd, Apt 505, Galveston, TX 77551',
 '8020 

In [291]:
# Verify the data from API source

# Filter API response to include only properties with a matched location
filtered_properties = [prop for prop in TX_final_combined if prop['formattedAddress'] in matches_2]

# Initialize a dictionary to hold the roomCount for each matched property
unit_counts = {}

# Extract unitCount for each filtered property
for prop in filtered_properties:
    # Extracting formattedAddress for identification
    address = prop['formattedAddress']
    # Checking if 'features' is present and if 'roomCount' is available within 'features'
    unit_count = prop.get('features', {}).get('unitCount', 'Not Specified')
    unit_counts[address] = unit_count

# Print unit counts for matched locations
for address, count in unit_counts.items():
    print(f"{address}: Unit Count = {count}")


4400 Ave N, Galveston, TX 77550: Unit Count = Not Specified
3428 Cove View Blvd, Galveston, TX 77554: Unit Count = Not Specified
4925 Fort Crockett Blvd, Galveston, TX 77551: Unit Count = Not Specified
1302 39th St, Galveston, TX 77550: Unit Count = Not Specified
8701 Teichman Rd, Galveston, TX 77554: Unit Count = Not Specified
1108 Ave L, Galveston, TX 77550: Unit Count = 1
3102 Cove View Blvd, Galveston, TX 77554: Unit Count = 8
7019 Lasker Dr, Galveston, TX 77551: Unit Count = Not Specified
1524 Strand St, Galveston, TX 77550: Unit Count = Not Specified
215 Market St, Galveston, TX 77550: Unit Count = Not Specified
3433 Cove View Blvd, Galveston, TX 77554: Unit Count = Not Specified
8020 Stewart Rd, Galveston, TX 77551: Unit Count = Not Specified
6424 Central City Blvd, Galveston, TX 77551: Unit Count = Not Specified
3219 69th St, Galveston, TX 77551: Unit Count = Not Specified
3222 69th St, Galveston, TX 77551: Unit Count = Not Specified
210 Strand St, Galveston, TX 77550: Unit Cou

In [294]:

# For roomCount:
filtered_properties = [prop for prop in TX_final_combined if prop['formattedAddress'] in matches_2]

# Initialize a dictionary to hold the roomCount for each matched property
unit_counts = {}

# Extract unitCount for each filtered property
for prop in filtered_properties:
    # Extracting formattedAddress for identification
    address = prop['formattedAddress']
    # Checking if 'features' is present and if 'roomCount' is available within 'features'
    unit_count = prop.get('features', {}).get('roomCount', 'Not Specified')
    unit_counts[address] = unit_count

# Print unit counts for matched locations
for address, count in unit_counts.items():
    print(f"{address}: room Count = {count}")


4400 Ave N, Galveston, TX 77550: room Count = Not Specified
3428 Cove View Blvd, Galveston, TX 77554: room Count = Not Specified
4925 Fort Crockett Blvd, Galveston, TX 77551: room Count = Not Specified
1302 39th St, Galveston, TX 77550: room Count = 10
8701 Teichman Rd, Galveston, TX 77554: room Count = Not Specified
1108 Ave L, Galveston, TX 77550: room Count = Not Specified
3102 Cove View Blvd, Galveston, TX 77554: room Count = Not Specified
7019 Lasker Dr, Galveston, TX 77551: room Count = Not Specified
1524 Strand St, Galveston, TX 77550: room Count = Not Specified
215 Market St, Galveston, TX 77550: room Count = Not Specified
3433 Cove View Blvd, Galveston, TX 77554: room Count = Not Specified
8020 Stewart Rd, Galveston, TX 77551: room Count = Not Specified
6424 Central City Blvd, Galveston, TX 77551: room Count = Not Specified
3219 69th St, Galveston, TX 77551: room Count = Not Specified
3222 69th St, Galveston, TX 77551: room Count = Not Specified
210 Strand St, Galveston, TX 77

## We see that 5 of the properties have unit count, and 3 of them have room count information.

Perhaps we cannot only rely on these two parameters. We might want to do some exploratory analysis on other parameters and use them to predict the size of residential buildings. Ohterwise, I might need to explore footprint datasets from Microsoft and U.S. Federal Emergency Management Authority (FEMA). 